Libraries/LibCore/EventLoop.cpp

```diff
@@ -409,6 +412,91 @@ void EventLoop::post_event(Object& receiver, NonnullOwnPtr<Event>&& event)
     m_queued_events.empend(receiver, move(event));
 }
 
+template <typename RemoveHandler>
+class ProtectSignalHandler
+{
+    AK_MAKE_NONCOPYABLE(ProtectSignalHandler);
+    AK_MAKE_NONMOVABLE(ProtectSignalHandler);
+
+    int& m_handling_signal;
+    int m_handling_signal_prev;
+    RemoveHandler m_remove_handler;
+
+public:
+    ProtectSignalHandler(int& handling_signal, int signal, RemoveHandler remove_handler):
+        m_handling_signal(handling_signal),
+        m_handling_signal_prev(handling_signal),
+        m_remove_handler(move(remove_handler))
+    {
+        handling_signal = signal;
+    }
+
+    ~ProtectSignalHandler()
+    {
+        bool was_removed = (m_handling_signal == 0);
+        m_handling_signal = m_handling_signal_prev;
+        if (was_removed) {
+            // This handler was flagged to be removed
+            m_remove_handler();
+        }
+    }
+};
+
+void EventLoop::handle_signal(int signo)
+{
+    // We need to protect the handler from being removed while handling it
+    ProtectSignalHandler protect(m_handling_signal, signo,
+        [&] {
+           ASSERT(m_handling_signal == signo); // should have been restored
+           unregister_signal(signo);
+        });
+    auto handler = m_signal_handlers.find(signo);
+    if (handler != m_signal_handlers.end()) {
+#ifdef EVENTLOOP_DEBUG
+        dbg() << "Core::EventLoop: handling signal " << signo;
+#endif
+        handler->value(signo);
+    }
+}
+
+void EventLoop::register_signal(int signo, Function<void(int)> handler)
+{
+    ASSERT(signo != 0);
+    ASSERT(m_handling_signal != signo); // can't re-register while handling it
+    if (m_signal_handlers.set(signo, move(handler))) {
+        signal(signo,
+            [](int sig) {
+                ASSERT(sig != 0);
+                // We MUST check if the current pid still matches, because there
+                // is a window between fork() and exec() where a signal delivered
+                // to our fork could be inadvertedly routed to the parent process!
+                if (getpid() == s_pid) {
+                    int nwritten = write(s_wake_pipe_fds[1], &sig, sizeof(sig));
+                    if (nwritten < 0) {
+                        perror("EventLoop::register_signal: write");
+                        ASSERT_NOT_REACHED();
+                    }
+                } else {
+                    // We're a fork who received a signal, reset s_pid
+                    s_pid = 0;
+                }
+            });
+    }
+}
+
+void EventLoop::unregister_signal(int signo)
+{
+    ASSERT(signo != 0);
+    if (m_handling_signal == signo) {
+        // We're removing the signal while handling it, flag removal
+        m_handling_signal = 0;
+    } else {
+        if (m_signal_handlers.remove(signo)) {
+            signal(signo, SIG_DFL);
```

Comment:
It might be slightly "correcter" to actually save the return value from `signal` and restore that? Is there a use case in the Shell for actually setting the signal mask for the signals it's trying to register handlers for? If so, exposing an API that looks more like `sigaction` would be the way to go. It is kind of sad that Serenity doesn't support the SA_SIGINFO flag yet, but that's just a wishlist item :)